internal/endpoints/awsrulesfn/arn.go

```diff
@@ -0,0 +1,120 @@
+package awsrulesfn
+
+import (
+	"fmt"
+	"strings"
+	smithyerrep "github.com/aws/smithy-go/error/endpoints"
+)
+
+// ARN provides AWS ARN components broken out into a data structure.
+type ARN struct {
+	Partition  string
+	Service    string
+	Region     string
+	AccountId  string
+	ResourceId OptionalStringSlice
+}
+
+const (
+	arnDelimiters      = ":"
+	resourceDelimiters = "/:"
+	arnSections        = 6
+	arnPrefix          = "arn:"
+
+	// zero-indexed
+	sectionPartition = 1
+	sectionService   = 2
+	sectionRegion    = 3
+	sectionAccountID = 4
+	sectionResource  = 5
+)
+
+// ParseARN returns an [ARN] value parsed from the input string provided. If
+// the ARN cannot be parsed nil will be returned, and error added to
+// [ErrorCollector].
+func ParseARN(input string, ec *smithyerrep.ErrorCollector) *ARN {
+	if !strings.HasPrefix(input, arnPrefix) {
+		ec.AddError(smithyerrep.FnError{
+			Name: "ParseARN",
+			Err:  fmt.Errorf("invalid ARN prefix, %q", input),
+		})
+		return nil
+	}
+
+	sections := strings.SplitN(input, arnDelimiters, arnSections)
+	if numSections := len(sections); numSections != arnSections {
+		ec.AddError(smithyerrep.FnError{
+			Name: "ParseARN",
+			Err: fmt.Errorf("invalid ARN, not enough sections, %q, %d",
+				input, numSections),
+		})
+		return nil
+	}
+
+	if sections[sectionPartition] == "" {
+		ec.AddError(smithyerrep.FnError{
+			Name: "ParseARN",
+			Err: fmt.Errorf("invalid ARN, partition section cannot be empty, %v",
+				input),
+		})
+		return nil
+	}
+	if sections[sectionService] == "" {
+		ec.AddError(smithyerrep.FnError{
+			Name: "ParseARN",
+			Err: fmt.Errorf("invalid ARN, service section cannot be empty, %v",
+				input),
+		})
+		return nil
+	}
+	if sections[sectionResource] == "" {
+		ec.AddError(smithyerrep.FnError{
+			Name: "ParseARN",
+			Err: fmt.Errorf("invalid ARN, resource section cannot be empty, %v",
+				input),
+		})
+		return nil
+	}
+
+	return &ARN{
+		Partition:  sections[sectionPartition],
+		Service:    sections[sectionService],
+		Region:     sections[sectionRegion],
+		AccountId:  sections[sectionAccountID],
+		ResourceId: splitResource(sections[sectionResource]),
+	}
+}
+
+// splitResource splits the resource components by the ARN resource delimiters.
+func splitResource(v string) []string {
+	var parts []string
+	var offset int
+
+	for offset <= len(v) {
+		idx := strings.IndexAny(v[offset:], "/:")
+		if idx < 0 {
+			parts = append(parts, v[offset:])
+			break
+		}
+		parts = append(parts, v[offset:idx+offset])
+		offset += idx + 1
+	}
+
+	return parts
+}
+
+// OptionalStringSlice provides a helper to safely get the index of a string
+// slice that may be out of bounds. Returns pointer to string if index is
+// valid. Otherwise returns nil.
+type OptionalStringSlice []string
```

Comment:
**Question**: I see `ResourceId` is set to this type but how is it used in practice? Why do we need the type to be `OptionalStringSlice`?

Reply:
I'm assuming the rules generator leverages the receiver method `Get` to simplify the codegen.

Reply:
Confirmed--looking at the SEP / S3 rules, the expectation is that you can make a `getAttr` call with an array index, and the return will be null if OOB. This comes into play in practice when cracking open resource IDs in arns.

(we got this for free in kotlin via list `getOrNull`)

Reply:
Yah, I believe Luc is right here. As stated in the SEP, the `getAttr` function is supposed to return a nullable field on indexes into the `resourceId` returned by `parseArn`. So instead of having the implementation of the `getAttr` function handle the Go error for index out of bounds, we contain that error in the `OptionalStringSlice` because as far as the SEP is concerned, it either returns the item or `nil`.